wallet/src/main/java/com/android/identity_credential/wallet/PresentationActivity.kt

```diff
@@ -379,18 +291,130 @@ class PresentationActivity : FragmentActivity() {
+        return PresentationRequestData(
+            document = credentialDocument,
+            documentRequest = credentialDocumentRequest,
+            docType = docRequestToUse.docType,
+            trustPoint = trustPoint,
+            sessionTranscript = deviceRetrievalHelper!!.sessionTranscript
+        )
     }
 
+
+    /**
+     * Return a credential identifier which can satisfy the request.
+     *
+     * If multiple credentials can satisfy the request, preference is given to the currently
+     * focused credential in the main pager.
+     *
+     * @param settingsModel app-wide settings
+     * @param docRequest the parsed [DeviceRequestParser.DocRequest]
+     * @return credential identifier if found, otherwise null.
+     */
+    private fun findFirstDocumentSatisfyingRequest(
+        settingsModel: SettingsModel,
+        docRequest: DeviceRequestParser.DocRequest,
+    ): String? {
+        // prefer the credential which is on-screen if possible
+        val credentialIdFromPager: String? = settingsModel.focusedCardId.value
+        if (credentialIdFromPager != null
+            && canDocumentSatisfyRequest(credentialIdFromPager, docRequest)
+        ) {
+            return credentialIdFromPager
+        }
+
+        return walletApp.documentStore.listDocuments().firstOrNull { credentialId ->
+            canDocumentSatisfyRequest(credentialId, docRequest)
+        }
+    }
+
+    /**
+     * Return whether the passed credential id can satisfy the request
+     *
+     * @param credentialId the id of the credential to check for satisfaction.
+     * @param docRequest the DocRequest with its DocType.
+     * @return whether the specified credential id can satisfy the request.
+     */
+    private fun canDocumentSatisfyRequest(
+        credentialId: String,
+        docRequest: DeviceRequestParser.DocRequest
+    ): Boolean {
+        val credential = walletApp.documentStore.lookupDocument(credentialId)!!
+        return credential.documentConfiguration.mdocConfiguration?.docType == docRequest.docType
+    }
+
+    /**
+     * Send response bytes to requesting party and updates state to [State.RESPONSE_SENT]
+     * @param deviceResponseBytes response bytes that may or may not have been processed (such as
+     * when sending an error)
+     */
+    private fun sendResponseToDevice(deviceResponseBytes: ByteArray) {
+        deviceRetrievalHelper?.sendDeviceResponse(
+            deviceResponseBytes,
+            Constants.SESSION_DATA_STATUS_SESSION_TERMINATION
+        )
+        state.value = State.RESPONSE_SENT
+    }
+
+
     private fun getDeviceRequest(): ByteArray {
         check(state.value == State.REQUEST_AVAILABLE) { "Not in REQUEST_AVAILABLE state" }
-        check(deviceRequest != null) { "No request available " }
-        return deviceRequest as ByteArray
+        check(deviceRequestByteArray != null) { "No request available " }
+        return deviceRequestByteArray as ByteArray
     }
```

Comment:
Why have this function in the first place, why would you call getDeviceRequest() instead of just accessing the `deviceRequestByteArray` property?